crypto/dilithium/p_pqdsa.c

```diff
@@ -0,0 +1,287 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/mem.h>
+
+#include "../crypto/internal.h"
+#include "../fipsmodule/evp/internal.h"
+#include "../evp_extra/internal.h"
+#include "ml_dsa.h"
+#include "internal.h"
+
+// PQDSA PKEY functions
+
+typedef struct {
+  const PQDSA *pqdsa;
+} PQDSA_PKEY_CTX;
+
+static int pkey_pqdsa_init(EVP_PKEY_CTX *ctx) {
+  PQDSA_PKEY_CTX *dctx;
+  dctx = OPENSSL_zalloc(sizeof(PQDSA_PKEY_CTX));
+  if (dctx == NULL) {
+    return 0;
+  }
+
+  ctx->data = dctx;
+
+  return 1;
+}
+
+static void pkey_pqdsa_cleanup(EVP_PKEY_CTX *ctx) {
+  OPENSSL_free(ctx->data);
+}
+
+static int pkey_pqdsa_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  GUARD_PTR(ctx);
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  GUARD_PTR(dctx);
+  const PQDSA *pqdsa = dctx->pqdsa;
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    pqdsa = PQDSA_KEY_get0_dsa(ctx->pkey->pkey.pqdsa_key);
+  }
+
+  PQDSA_KEY *key = PQDSA_KEY_new();
+  if (key == NULL ||
+      !PQDSA_KEY_init(key, pqdsa) ||
+      !pqdsa->method->keygen(key->public_key, key->secret_key) ||
+      !EVP_PKEY_assign_PQDSA_KEY(pkey, key)) {
+    PQDSA_KEY_free(key);
+    return 0;
+      }
+  return 1;
+}
+
+static int pkey_pqdsa_sign_signature(EVP_PKEY_CTX *ctx, uint8_t *sig,
+                                     size_t *siglen, const uint8_t *tbs,
+                                     size_t tbslen) {
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  const PQDSA *pqdsa = dctx->pqdsa;
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    pqdsa = PQDSA_KEY_get0_dsa(ctx->pkey->pkey.pqdsa_key);
+  }
+
+  // Caller is getting parameter values.
+  if (sig == NULL) {
+    if (pqdsa != NULL) {
```

Comment:
unnecessary, already checked in line 65. `siglen` should be checked before writing to it.